aggregator/election_mgr.go

```diff
@@ -0,0 +1,545 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/leader/campaign"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/watch"
+
+	"github.com/uber-go/tally"
+)
+
+// ElectionManager manages leadership elections.
+type ElectionManager interface {
+	// Open opens the election manager for a given shard set.
+	Open(shardSetID string) error
+
+	// ElectionState returns the election state.
+	ElectionState() ElectionState
+
+	// Resign stops the election and resigns from the ongoing campaign if any, thereby
+	// forcing the current instance to become a follower. If the provided context
+	// expires before resignation is complete, the context error is returned, and the
+	// election is restarted if necessary.
+	Resign(ctx context.Context) error
+
+	// Close the election manager.
+	Close() error
+}
+
+const (
+	backOffOnResignOrElectionError = time.Second
+)
+
+var (
+	errElectionManagerAlreadyOpenOrClosed = errors.New("election manager is already open or closed")
+	errElectionManagerNotOpenOrClosed     = errors.New("election manager is not open or closed")
+	errLeaderNotChanged                   = errors.New("leader has not changed")
+	errResignActionAlreadyEnqueued        = errors.New("resign action already enqueued")
+)
+
+type electionManagerState int
+
+const (
+	electionManagerNotOpen electionManagerState = iota
```

Comment:
So `NotOpen`, `Open` and `Closed` are the three states we've consistently used in all of our repos. Technically speaking `NotOpen` is also a state. I'd prefer to stick to the convention, but can change if you feel strongly about it.